Add rack/dc aware load balancing

dkropachev opened the pull request:
Closes https://github.com/scylladb/alternator-load-balancing/issues/35

Core PR that introduced rack and dc filtering on `/localnodes`: https://github.com/scylladb/scylladb/issues/12147

This PR allows user to target dynamodb client to a particular rack/datacenter or rack+datacenter, via following API:
```
        AlternatorRequestHandler handler = new AlternatorRequestHandler(uri, datacenter, rack);
```

Changes in auxiliary `AlternatorLiveNodes` API:
1. Constructor input argument is an single URI now
2. Validation is introduced
3. Start is moved out from the constructor 
```
        liveNodes = new AlternatorLiveNodes(Collections.singletonList(seedURI), datacenter, rack);
        try {
            liveNodes.validate();
            liveNodes.checkIfRackAndDatacenterSetCorrectly();
        } catch (AlternatorLiveNodes.ValidationError | AlternatorLiveNodes.FailedToCheck e) {
            throw new RuntimeException(e);
        }
        liveNodes.start();
```

### Tested
1. Regular, old code: `AlternatorRequestHandler handler = new AlternatorRequestHandler(uri)` for `6.2.0` and `master`
2. Rack-aware: `AlternatorRequestHandler handler = new AlternatorRequestHandler(uri, "dc1", "rack1")`  for `6.2.0` and `master`
3. Dc-aware: `AlternatorRequestHandler handler = new AlternatorRequestHandler(uri, "dc1", "")`  for `6.2.0` and `master`

All tests are done for following deployments:
1. MultiDc
2. MultiRack+MultiDc
3. SingleDc

mykaul commented:
@nyh please take a look.

nyh commented:
If I understand correctly, the example you gave and the Demo1 you modified are for the old Java SDK (v1). Did you also check the newer SDK v2 (it was released 7 years ago - it isn't very new...)? If you didn't, please do. If you did, please mention that both are supported (or not) in the commit message.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -86,14 +91,101 @@ public URI nextAsURI() {
+        String path = "/localnodes";
+        boolean paramsInitiated = false;
+        if (this.isRackSpecified()) {
+            path += "?rack=" + this.rack;
```
nitpick: you may need to URL-encode the rack and dc names if they contain strange characters. Maybe not worth the hassle, but we may be suprised in the future if a customer uses strange characters in a rack/dc name...

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -86,14 +91,101 @@ public URI nextAsURI() {
+        return new URL(alternatorScheme, nextNode(), alternatorPort, file);
+    }
+
+    public RuntimeException checkIfRackAndDatacenterSetCorrectly() {
```
What does it mean for this method to return - not throw - an exception?

dkropachev replied:
Yes, initially i was thinking that i can reuse this part of the code and fotgot to check on it later, switched to throwing them

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorRequestHandler.java:
```diff
@@ -5,19 +5,53 @@
+        return this;
+    }
+
+    public AlternatorRequestHandler checkIfRackAndDatacenterSetCorrectly() throws RuntimeException {
```
Please add documentation on what exactly this function checks about the setting being "correct".
If I understand correctly (but please correct me if I'm wrong), it doesn't really check if the rack and datacenter are "correct" but just if this feature is supported in the server. In this case I have several thoughts/questions:
1. Why bother check - why would a user try to set a rack if the server doesn't support it?
2. If a user does want to check, why not have a function called checkSupportsRackandDatacenter which the user can call? Why would the user call this after already setting datacenter and rack?
3. But on the other hand, why does the user need to call this explicitly - setrack()/setdc() could check immediately.

dkropachev replied:
1. We have had exactly same problem for drivers, user missconfigure driver and then does not know why it does not work or behaive in a way it does not expect.
2. I will do that.
3. Agree setRack, is not needed at this point, i will remove it.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -103,36 +195,78 @@ private static String streamToString(java.io.InputStream is) {
+        List<String> hostsWithFakeRack = getNodes(fakeRackUrl);
+        List<String> hostsWithoutRack = getNodes(url);
+        if (hostsWithoutRack.isEmpty()) {
+            // If list of nodes is empty, it is impossible to conclude if it supports rack/datacenter filtering or not.
```
This case may have some theoretical interest, but it practice it should not be possible and indicates a logic error in Scylla. The whole assumption of this load balancer is that we start with one known live node, and can contact it to get the list of all live nodes. If a node responds but says it itself isn't live, nothing will work.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -86,14 +91,101 @@ public URI nextAsURI() {
+    private boolean isRackAndDatacenterSupported() {
+        if (this.supportsRackAndDatacenter == null) {
+            LocalDateTime now = LocalDateTime.now();
+            if (this.nextRackAndDatacenterFeatureSniffTime == null || now.isAfter(this.nextRackAndDatacenterFeatureSniffTime)) {
```
Why do we need to recheck every minute that the dc/rack feature is still supported?

dkropachev replied:
I am going to drop it.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -103,36 +195,78 @@ private static String streamToString(java.io.InputStream is) {
+        Collections.sort(hostsWithFakeRack);
+        Collections.sort(hostsWithoutRack);
+        if (hostsWithoutRack.equals(hostsWithFakeRack)) {
+            // When rack filtering is not supported server returns same nodes.
```
You could have simplified this check - if rack filtering is supported, the result of filtering with a fake rack name is always an empty list. No real need to sort or compare hostsWithoutRack.

dkropachev replied:
It is just protection against the case when `fakeRack` is there.

nyh replied:
You mean when there is an actual rack called "fakeRack"? In this case, this check is wrong anyway - it will return "false" (because of the equality) even when the rack option is supported.
To avoid the case where fakeRack is actually exists, you can use a longer ridiculous name like `this_FAKE_rack_couldn't possibly exist!` ;-) Of course that's not perfect either.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorRequestHandler.java:
```diff
@@ -5,19 +5,53 @@
+        return this;
+    }
+
+    public AlternatorRequestHandler start() {
```
This changes the API of AlternatorRequestHandler - not only did you add setDataCenter()/setRack() methods, you also made it necessary to call the start() method - something that wasn't needed in the past.
This will cause problems to existing users of this code (if we have any...), but moreover - the README.md which explains how to use this will need to be changed to call this start().

Personally, I think it would be better not to introduce this start() - I would just make the datacenter/rack an (optional) constructor parameter, just like uri already is. Why would that not be better?

dkropachev replied:
Well, my concern is that when class instance invokes threading, opens a file, opens a network connection during initialization it makes it hard and hacky to test such classes.

Alternative is to implement `AlternatorLiveNodes` with `start` on it, while in `AlternatorRequestHandler` and `AlternatorEndpointProvider` call `start` in a constructor.

nyh replied:
I don't understand why the _implementation_ of a class (whether it relies on some background thread or not) makes it easier or harder to test, but my point was that it wasn't a **necessary** part of this patch, which had other goals.

But perhaps more important I commented that if you make this API change you also need to:
1. Change the documentation - not just demo1.java - to explain the new way of creating this class.
2. Change not just the SDK v1 class and demo1.java, but also the SDK v2 class and demo2.java, because those use the changed API as well.

Your new suggestion (having start() in the internal AlternatorLiveNodes but not in the external class) makes 1 above unnecessary, but 2 still is. But it still makes me wonder what you gain by this internal change, which just adds confusion (now there is an internal API which is different from the external API).

To make it clear, I'm not strongly against this change, only mildly against it (until I hear a convincing example of why the thread-in-constructor thing is problematic), but do insist that if you do this change you should do everything that it needs - including changing the README.md and the SDK v2 code to match the change.

roydahan replied:
@dkropachev I recommend not doing unnecessary changes at the moment.

nyh reviewed (changes requested):
Thanks, looks mostly good for SDK v1 (I think it breaks v2?), but I think but I have one objection about the need for a new start() which needlessly (I think) changes the API of this library. Wouldn't it be easier if we just had a constructor that takes in addition to a url, also dc and rack? Basically, exactly the same code you added to Demo1 (the constructor will take both dc and rack, but null means not to set one or both).

I am also not sure there is a need for all the complexity you added for checking if the rack/dc feature is supported, and retesting this every minute. After all, the rack/dc option is something the user explicitly chooses for a specific cluster - just like the initial address of one known node. Why would a user pass wrong parameters? Why does the library need to check and recheck it, instead of just saying - if you pass wrong parameters, you'll get wrong results?

I don't mind if we have a method like checkSupportForRackandDC() or something like that, maybe some user will find it convenient, but this can be a stand-alone function and doesn't need to be mixed into all the other functions.

dkropachev commented:
> Thanks, looks mostly good for SDK v1 (I think it breaks v2?), but I think but I have one objection about the need for a new start() which needlessly (I think) changes the API of this library. Wouldn't it be easier if we just had a constructor that takes in addition to a url, also dc and rack? Basically, exactly the same code you added to Demo1 (the constructor will take both dc and rack, but null means not to set one or both).

I consider having thread running when you initialize class as bad practicce that does not allow properly test it and makes code hard to read and understand:
In this particular case if you did not look at constructor you would not even know that there is a thread running underneeth. 

> I am also not sure there is a need for all the complexity you added for checking if the rack/dc feature is supported, and retesting this every minute. After all, the rack/dc option is something the user explicitly chooses for a specific cluster - just like the initial address of one known node. Why would a user pass wrong parameters? Why does the library need to check and recheck it, instead of just saying - if you pass wrong parameters, you'll get wrong results?

We had exactly same problem in drivers, user provides wrong rack, datacenter or something and can't figure out why queries are failing.
With this code they have opportunity to explicitly check if they match.

Regarding check if feature is supported or not. same reason, if we can tell user that target cluster does not support feature, we should do that, otherwise code will produce unexpected results and user left to guess what is going on.

> I don't mind if we have a method like checkSupportForRackandDC() or something like that, maybe some user will find it convenient, but this can be a stand-alone function and doesn't need to be mixed into all the other functions.

Makes sense, let's do it this way.

nyh commented:
> > Thanks, looks mostly good for SDK v1 (I think it breaks v2?), but I think but I have one objection about the need for a new start() which needlessly (I think) changes the API of this library. Wouldn't it be easier if we just had a constructor that takes in addition to a url, also dc and rack? Basically, exactly the same code you added to Demo1 (the constructor will take both dc and rack, but null means not to set one or both).
> 
> I consider having thread running when you initialize class as bad practicce that does not allow properly test it and makes code hard to read and understand: In this particular case if you did not look at constructor you would not even know that there is a thread running underneeth.

I don't why the user needs to care if the object contains a thread or not, but if you decided to change this part (and you didn't strictly have to do this, you could have added a constructor parameter), you'll need to change the documentation - and hope nobody is using the existing version and will complain about the API change :-)
Also I think the SDK v2 also uses this class so will need to change too.


> > I am also not sure there is a need for all the complexity you added for checking if the rack/dc feature is supported, and retesting this every minute. After all, the rack/dc option is something the user explicitly chooses for a specific cluster - just like the initial address of one known node. Why would a user pass wrong parameters? Why does the library need to check and recheck it, instead of just saying - if you pass wrong parameters, you'll get wrong results?
> 
> We had exactly same problem in drivers, user provides wrong rack, datacenter or something and can't figure out why queries are failing. With this code they have opportunity to explicitly check if they match.
> Regarding check if feature is supported or not. same reason, if we can tell user that target cluster does not support feature, we should do that, otherwise code will produce unexpected results and user left to guess what is going on.

So the setrack() function (or the constructor, in the old style) can do the check. Why would a user even think to call a separate function to check this? And why does this feature need to be re-checked every minute - are we worried that the user will downgrade the cluster and lose this feature (I don't even think this is possible in Scylla)?

Bouncheck commented on java/src/test/java/com/scylladb/alternator/test/Demo1.java:
```diff
@@ -60,7 +60,29 @@ static DynamoDB getTraditionalClient(URI url) {
+        if (datacenter != null && !datacenter.isEmpty()) {
+            handler.setDatacenter(datacenter);
+        }
+        handler.checkIfRackAndDatacenterSetCorrectly().start();
```
I think it should not be needed to manually call checks for various options before start. Validation of all options could be done on start.

dkropachev replied:
Done.

Bouncheck reviewed (commented):
I've added two small comments, since other important things are already covered by other reviewers.

dkropachev commented:
@nyh , @Bouncheck , let's split it into parts, i will create smaller PRs to address separate issues and then we proceed with this one.

nyh commented:
> @nyh , @Bouncheck , let's split it into parts, i will create smaller PRs to address separate issues and then we proceed with this one.

Yes, although note that the rack-aware load balancing is the most important part, I think it should be done first before all the other minor (?) improvements. Also, I asked why you decided to fix this only for the v1 version of the SDK and not v2, and didn't get a reply.

dkropachev commented:
> > @nyh , @Bouncheck , let's split it into parts, i will create smaller PRs to address separate issues and then we proceed with this one.
> 
> Yes, although note that the rack-aware load balancing is the most important part, I think it should be done first before all the other minor (?) improvements. Also, I asked why you decided to fix this only for the v1 version of the SDK and not v2, and didn't get a reply.

@nyh, @Bouncheck , could you please take a look agian, I have addressed comments and rebased to the master, also I have fixed both `Demo1`, `Demo2` and `Demo3`

nyh commented:
@dkropachev in a single-patch pull-request, it is the single patch's commit message - NOT the cover letter - which gets inserted into the git history and there will be no merge and therefore no cover letter. So please make sure that this single patch's commit message has the full description - and not just a single line.

nyh commented:
> @dkropachev in a single-patch pull-request, it is the single patch's commit message - NOT the cover letter - which gets inserted into the git history and there will be no merge and therefore no cover letter. So please make sure that this single patch's commit message has the full description - and not just a single line.

Oh, and of course please ensure that the commit message is up-to-date with the code. I see for example that instead of the setrack() et al. that you described, now it's part of the constructor.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorEndpointProvider.java:
```diff
@@ -41,4 +47,8 @@ public CompletableFuture<Endpoint> resolveEndpoint(DynamoDbEndpointParams endpoi
 		return endpoint;
 	}
+
+	public void checkIfRackAndDatacenterSetCorrectly() throws AlternatorLiveNodes.FailedToCheck, AlternatorLiveNodes.ValidationError {
```
If you have such a method, I think it needs a comment - what does it mean for a rack and datacenter to be set "correctly"?
If I understand correctly, what you really want to check is: **If** a specific datacenter and/or rack were requested, this function verifies that the server supports the server and rack option. But it does NOT verify (again, if I understand correctly) if the specific rack/datacenter chosen actually exists.

Also, we discussed this earlier - why does there need to be a separate function for this, instead of the constructor checking this if datacenter/rack is set?

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -1,15 +1,12 @@
 
 import java.io.IOException;
-import java.net.MalformedURLException;
+import java.net.*;
```
Nitpick (feel free to reject): I think it is considered "not good" to use stars in Java import, and you should make an effort to list the specific imports you need. Why did you want to replace the four imports with this star? To avoid adding a fifth import? But why did you want to avoid it?

dkropachev replied:
It is just IDE magic, will roll it back

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -55,48 +53,57 @@ public void run() {
-    public AlternatorLiveNodes(String alternatorScheme, List<String> liveNodes, int alternatorPort) {
-        this.alternatorScheme = alternatorScheme;
-        this.initialNodes = new ArrayList<>(liveNodes);
+    public AlternatorLiveNodes(List<URI> liveNodes, String datacenter, String rack) {
```
I don't like this change to the constructor... I mean the change to use `List<URI>`:
The existing interface assumed that we have a list of node hostnames (each a String), and all of them share the same scheme and port. This assumption is a valid one because this is indeed how Scylla works (all nodes use the same scheme and port), and because "/localnodes" returns just the host names, not the port and scheme, and we need to know what it will be for all nodes.

Now you are changing this constructor to take a list of URIs. Theoretically, a caller can pass in each one a different port and scheme, but if you try to do that, nothing will work! In fact, your code saves the port and scheme of the **first** node of the list, and then assumes (without checking!) that all the rest have the same port and scheme. But if it's what you do, why even allow passing separate URIs?

I think you shouldn't change the original interface with the separate scheme, port and hostname. Internally, if you want to build a list of URIs from that, you can (I think it's a bit wasteful, but not that much) - but it shouldn't be the interface.

dkropachev replied:
Upper API converts string to URI and then converts it back to String `uri.getHost()`, which is a bit weird.
Is it ok to have it as `public AlternatorLiveNodes(URI liveNode, String datacenter, String rack)` ?

dkropachev replied:
Good point, fixed.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +150,125 @@ private static String streamToString(java.io.InputStream is) {
-            // Should never happen, uri is already validated at this point
-            throw new RuntimeException(e);
+            conn.setRequestMethod("GET");
+        } catch (ProtocolException ignored) {
```
Please add a comment what this catch is about. I assume it should never happen because the URI was already "validated" earlier?

dkropachev replied:
Added a comment, please take a look.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +150,125 @@ private static String streamToString(java.io.InputStream is) {
+                query += "&dc=" + this.datacenter;
+            }
+        }
+        return nextAsURI("/localnodes", query);
```
Now I'm really puzzled. I thought the whole idea of using URI instead of the strings we worked with in the past was that the URI somehow made it easier for us to correctly build this query string. But now I see we just build it as a string exactly as before. So what did we gain?

dkropachev replied:
This part is called once a second, it is not high loaded part of the code.
What we wanted to avoid is convert String to URI every time `AlternatorEndpointProvider` and `AlternatorRequestHandler` need a next node and at the same time having nodes validated when the list is loaded from an API.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +150,125 @@ private static String streamToString(java.io.InputStream is) {
+    }
+
+    /**
+     * Checks if datacenter and/or rack are set correctly.
```
This comment doesn't explain what this function checks - it just repeats the name of the function with spaces and doesn't explain at all what "correctness" is being checked. See my comment above what I would have liked such a comment to explain.

dkropachev replied:
Thanks, doc string is fixed.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +150,125 @@ private static String streamToString(java.io.InputStream is) {
+        try {
+            List<URI> nodes = getNodes(nextAsLocalNodesURI());
+            if (nodes.isEmpty()) {
+                throw new ValidationError("node returned empty list, datacenter or rack are set incorrectly");
```
Oh, I didn't realize you check that. IIRC you didn't check that in the previous version. All the more it is important to say in the comment that this is what this function checks.

dkropachev replied:
Done, please take a look at the comment.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +150,125 @@ private static String streamToString(java.io.InputStream is) {
+            List<URI> hostsWithFakeRack = getNodes(fakeRackUrl);
+            List<URI> hostsWithoutRack = getNodes(uri);
+            if (hostsWithoutRack.isEmpty()) {
+                // If list of nodes is empty, it is impossible to conclude if it supports rack/datacenter filtering or not.
```
I still think this comment should explain that this case can't happen - we can't be contacting a live node (this is the assumption) and then getting a message from it that the whole cluster has no live nodes.

dkropachev replied:
added it there.

nyh commented on java/src/test/java/com/scylladb/alternator/test/Demo1.java:
```diff
@@ -82,14 +90,53 @@ public static void main(String[] args) {
         logger.addHandler(handler);
         logger.setUseParentHandlers(false);
 
+        ArgumentParser parser = ArgumentParsers.newFor("Demo3").build()
```
This is Demo1, not Demo3.
Please seriously reconsider why you want to introduce so many lines and a non-standard library I never heard of - net.sourceforge.argparse4j - just for a demo of how to use the library. A demo is supposed to be a short example, not something very user-friendly.
If you do feel strongly about adding this command-line parsing, can you please consider using something built-in into Java and not some uncommon external library?

nyh replied:
I see now that @elcallio introduced this external library and the argument and made the demo code lengthier with parsing code in Demo3, in commit 27cf1115. This means I probably can't complain when you now copied it into Demo1 and Demo2. I still don't like it in any of the demo codes, but my case has become weaker...

dkropachev replied:
It is just copy-paste from the `Demo3.java`, I agree it is better to use std library, but since we already have dependancy on it, it is not big of the problem, can we have an issue to switch from `net.sourceforge.argparse4j` to something else ?

Reason why it is needed it makes testing code so much easier.
As of now I am talking about only local testing, but in future we will do integration tests to make sure all our demos work flawlessly and with expected result.

nyh commented on java/src/test/java/com/scylladb/alternator/test/Demo1.java:
```diff
@@ -82,14 +90,53 @@ public static void main(String[] args) {
+        parser.addArgument("-p", "--password").setDefault("none")
+                .help("Credentials password");
+        parser.addArgument("--datacenter").type(String.class).setDefault("")
+                .help("Target only nodes from particular datacenter");
```
I think such an option needs to be explained better. The load balancer **always** targets just one specific data center - with and without this option. The difference is how this data center is chosen - if a data center option isn't specified, this single data center is the one that this first node belongs to. If the data center is specific - that data center is used (and it doesn't matter if the first node is in a different data center).

nyh commented on java/src/test/java/com/scylladb/alternator/test/Demo1.java:
```diff
@@ -82,14 +90,53 @@ public static void main(String[] args) {
+        AWSCredentialsProvider myCredentials = new AWSStaticCredentialsProvider(new BasicAWSCredentials(user, pass));
+        DynamoDB ddb;
+        if (endpoint.isEmpty()) {
+            ddb = getTraditionalClient(URI.create(endpoint), myCredentials);
```
I don't understand this case... The example of getTraditionalClient also had a URL. Why do you call it only if the URL is empy? What does this even mean?

dkropachev replied:
No, in traditional case endpoint is being calculated by SDK from the region, while in case of alternator we override endpoint decision to inject sylla node in there.

nyh replied:
In this demo code, the "traditional client" was a traditional **client** on Alternator, i.e., it does need to know what is the endpoint. Then we have the load-balancing client.

nyh replied:
Also, I still don't understand the code that first checks if endpoint.isEmpty(), and if it is, it then calls URI.create(endpoint). I think that's just a bug!

dkropachev replied:
True, but it is, fixed.

dkropachev commented on java/src/main/java/com/scylladb/alternator/AlternatorEndpointProvider.java:
```diff
@@ -19,8 +21,12 @@ public class AlternatorEndpointProvider implements DynamoDbEndpointProvider {
 	private final Map<URI, CompletableFuture<Endpoint>> futureCache;
 
 	public AlternatorEndpointProvider(URI seedURI) {
+		this(seedURI, "", "'");
```
Thanks, nice catch, probably misscliked.

Bouncheck commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +150,125 @@ private static String streamToString(java.io.InputStream is) {
+            fakeRackUrl = new URI(uri.getScheme(), null, uri.getHost(), uri.getPort(), uri.getQuery(), "rack=fakeRack", "");
+        } catch (URISyntaxException e) {
+            // Should not ever happen
+            throw new FailedToCheck("Invalid URI: " + uri, e);
```
If this should not ever happen then i believe unchecked exception would be more appropriate. This forces the caller of the function to handle this case. If this case does not happen I think there is no reason to force handling of that.

Bouncheck replied:
Although there are other throws with this exception so the caller would need to do handle that anyway. I leave this to your judgement then.

dkropachev replied:
The idea there is that since there is already `FailedToCheck` present in the same method, which signals that it can't complete check properly it is better to use it, otherwise I would use `RuntimeException`

Bouncheck reviewed (commented):
Added one comment. Other things I saw were already covered.

dkropachev commented:
> > @dkropachev in a single-patch pull-request, it is the single patch's commit message - NOT the cover letter - which gets inserted into the git history and there will be no merge and therefore no cover letter. So please make sure that this single patch's commit message has the full description - and not just a single line.
> 
> Oh, and of course please ensure that the commit message is up-to-date with the code. I see for example that instead of the setrack() et al. that you described, now it's part of the constructor.

Done, please take a look at commit and PR description.

nyh commented:
> @nyh can you please prioritize this review so we can "release" it?

Will review it now.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +154,123 @@ private static String streamToString(java.io.InputStream is) {
+        if (this.rack.isEmpty() && this.datacenter.isEmpty()) {
+            return;
+        }
+        try {
```
I think you forgot to call checkIfRackDatacenterFeatureIsSupported() at this point?
You don't seem to call checkIfRackDatacenterFeatureIsSupported() in the code anywhere, and it's a private method, so it seems like a mistake.

dkropachev replied:
Added it to the handlers.

nyh reviewed (changes requested):
Hi, looks mostly good, but I still have a couple of things I'd like you to please fix (and a few more comments that I don't insist on). Please take a look at all the live comment threads on your single commit.

dkropachev commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -132,7 +139,7 @@ public URI nextAsURI() {
     public URI nextAsURI(String file, String query) {
```
Fixed.

nyh replied:
I don't see the fix. Maybe you forgot to push it?

dkropachev replied:
github UI does not pick up new commits on the branch, I am trying to fix it.

dkropachev commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +154,123 @@ private static String streamToString(java.io.InputStream is) {
+            try {
+                newHosts.add(this.hostToURI(host));
+            } catch (URISyntaxException | MalformedURLException e) {
+                logger.log(Level.WARNING, "Invalid host: " + host, e);
```
This is just to have no blank spots in validation to make sure that only kasher hosts made they way into host list, and not to handle this error later.
In general, it could be anything "/", first digits, symbols.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorEndpointProvider.java:
```diff
@@ -20,7 +20,13 @@ public class AlternatorEndpointProvider implements DynamoDbEndpointProvider {
 
 	public AlternatorEndpointProvider(URI seedURI) {
 		futureCache = new ConcurrentHashMap<>();
```
I forgot we merged this futureCache. I still don't understand why it's correct, or optimizing anything :-( But it's irrelevant to this patch.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorEndpointProvider.java:
```diff
@@ -20,7 +20,13 @@ public class AlternatorEndpointProvider implements DynamoDbEndpointProvider {
+		} catch (AlternatorLiveNodes.ValidationError e) {
+			throw new RuntimeException(e);
+		}
+		liveNodes.start();
```
Nitpick: (not important, will not insist on this)
It seems this code snippet - calling the constructor, then validate, then start - was the only correct way to create this object, which is why the create() factory function existed. I don't understand why you wanted to get rid of it in this patch.

dkropachev replied:
For now yes, in future, when we address https://github.com/scylladb/alternator-load-balancing/issues/46, having it `started` is going to be optional.
As I mentioned before, if we ever made it to unit-test, it will be way easier to have it tested with this API, then when everything is done on initialization.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -26,69 +27,115 @@ public class AlternatorLiveNodes extends Thread {
+                throw new RuntimeException(e);
+            }
+        }
+        this.liveNodes.set(nodes);
```
I still don't understand why this code - converting initialNodes into liveNodes - isn't in the constructor. If you did that, you wouldn't need the silly used-once class field "initialNodes", and you wouldn't need to do exactly the same check in validate().

I mentioned this already and it seems I didn't convince you - so I just want to mention this one last time and I won't insist - the feature is more important than perfecting those unrelated changes.

dkropachev replied:
This place was changed, but by some reason github UI does not pick up new commits on the branch, I am trying to fix it.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -26,69 +27,115 @@ public class AlternatorLiveNodes extends Thread {
+            try {
+                this.hostToURI(liveNode);
+            } catch (MalformedURLException | URISyntaxException e) {
+                throw new ValidationError(String.format("failed to validate initial node %s", liveNode), e);
```
This message isn't helpful for the user who'll see it. What does it mean that we "failed to validate" a node? It seems you only check that it's a valid-looking URL - not that it, for example, is a live node. You should say what sort of validation it failed.

dkropachev replied:
it will be included in the original exception, so they will see what exactly is wrong.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -26,69 +27,115 @@ public class AlternatorLiveNodes extends Thread {
-        } catch (URISyntaxException e) {
-            logger.log(Level.WARNING, "nextAsURI", e);
-            return null;
+            this.hostToURI("1.1.1.1");
```
I don't know what you were trying to do here, but I doubt this 1.1.1.1 thing was intentional :-(

dkropachev replied:
it is, I just want to make sure here that `alternatorScheme` and `alternatorPort` are good.

nyh replied:
Please add a comment that you are checking that the scheme and port are valid. I had no idea what you're doing here. It looked like a typo :-)

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -55,48 +54,57 @@ public void run() {
-        this.liveNodes = new AtomicReference<>();
-        this.alternatorPort = alternatorPort;
-        this.nextLiveNodeIndex = new AtomicInteger(0);
+    public AlternatorLiveNodes(URI liveNode, String datacenter, String rack) {
```
I don't understand why you are changing the API of this constructor again, after the previous patch already changed it.

Unlike the previous version of this pull request, the API you are suggesting here is not wrong (a single "liveNode" has just one port and scheme), but in my opinion, it's not better than the one before this patch - which was more general (allowed to add more than one initial node as the initial guess). I think it's a mistake to change it - it's not improving this class, it's making it less powerful.

dkropachev replied:
It addresses your prior [comment](https://github.com/scylladb/alternator-load-balancing/pull/40#discussion_r1855455731) in the following manner:
1. It allows user to create an instance with only one initial nodes and reads schema and port from it.
2. It also allows user to create an instance with multuple initial nodes, but in such case it does not assume schema and port from the first record, instead it forces user to pick correct schema and port for it.

nyh replied:
Yes, I agreed that this version is better than your previous change - at least it's not wrong - but I don't understand why you are changing this API.
Your point 2 seems wrong - after this change it is no longer possible to create an instance with multiple initial nodes, you can only specify one `URI liveNode`. If this is what we had all along it wouldn't have been a disaster, but before this patch we could specify a list of hostnames (strings) - how is this patch improving anything?

nyh commented on java/src/test/java/com/scylladb/alternator/test/Demo1.java:
```diff
@@ -1,5 +1,7 @@
 package com.scylladb.alternator.test;
 
+import com.amazonaws.services.dynamodbv2.xspec.B;
```
Is this a mistake (e.g., you typed B and pressed control-space) and got this?

nyh reviewed (changes requested):
I have a few minor objections to some of the improvements you made to the API and structure of this code, which I'm not sure improve anything, but I will NOT insist on any of them - I just want this patch to go in.

However, there is one last thing that seems **wrong** - the reference to "1.1.1.1" - either it's a typo or I don't understand what you tried to do there. This is why I'm chosing "Request changes" here. When you fix this error (?), we can merge this PR. You can think about the rest of my comments about changes you did and I didn't like, later :-)

dkropachev commented:
@nyh, there was an issue with github UI not picking up new commits from the branch, I managed to fix it by pushing whole branch, as it is to the `origin`.
Unfortunately, I did not spot it in time and last review your did was done on an outdated version.
I am sorry for it to happen, but I have to ask you to rereview it again.

nyh commented:
I'll review this again now.

nyh commented on java/src/main/java/com/scylladb/alternator/AlternatorLiveNodes.java:
```diff
@@ -147,38 +152,123 @@ private static String streamToString(java.io.InputStream is) {
+            conn.setRequestMethod("GET");
+        } catch (ProtocolException e) {
+            // It can happen only of conn is already connected or "GET" is not a valid method
+            // Both cases not true, os it should happen
```
**so** it should **not** happen.

nyh reviewed (approved):
I still have some comments about the code and about some of the changes to the existing code which I found unnecessary and perhaps even a change for the worse, but I'm still clicking "approve" and will merge this PR because I want this feature in, and if I understand none of the changes in this version are completely wrong, and most importantly: Because this code keeps changing every time I review it, and I want to merge something and then you can continue making changes to it. Please consider the comments I left in this PR - even after it's merged - as ideas for your continuing changes to this code.

nyh commented:
Despite I still have concerns and comments (that I'd like you to please consider in the future), I merged this PR, I think it's already good enough to be used.

One thing your patch doesn't change is the documentation java/README.md. The patch does change the demo codes (we have 3 versions of that...) to include the new feature, but not the document - so please consider changing the document too.

Namely, the document should explain the existance of the dc and rack options, and the possibilities it opens up, in particular:

1. The ability to do rack-aware load balancing (the documentation should also explain why it's a good idea - basically to save money on inter-rack communication).
2. The ability to find Scylla nodes on one DC even if the initial node you know about is actually in a different DC.

I don't think users need to guess that these feature exist, so the documentation should be updated.
